### PR TITLE
[Performance]Reduce sampling

### PR DESCRIPTION
**Pull request by hzx55906**
What this PR does / why we need it?
This PR introduces a significant optimization to the sampling process in a distributed (Tensor Parallel) environment, named "reduce sampling". The core goal is to minimize communication overhead during token selection, thereby improving the performance of vLLM Ascend in distributed deployment scenarios.

In traditional distributed sampling, full vocabulary gathering across tensor parallel ranks leads to high communication costs, especially in large-scale parallel settings. This optimization addresses the issue by designing specialized sampling logic that intelligently aggregates and processes vocabulary information, reducing unnecessary data transmission while ensuring sampling accuracy.

Does this PR introduce any user-facing change?

No breaking user-facing changes. A new configuration option enable_reduce_sample is added to AscendConfig, which users can optionally enable to activate the optimized distributed sampling scheme. Existing sampling behaviors remain unchanged by default.

How was this patch tested?

Tested with vLLM version based on v0.20.1. 
- Validation of distributed greedy sampling logic, ensuring correct global argmax selection across tensor parallel ranks.

- Verification of distributed top-k/top-p sampling, confirming accurate candidate aggregation and filtering.

- Functional testing of speculative decoding rejection sampling with the new compressed vocabulary mode.

- End-to-end testing of the integrated sampling pipeline in ModelRunnerV1, covering both regular and speculative decoding workflows.

Summary of Changes

- Optimized Distributed Sampling Scheme: Added enable_reduce_sample in AscendConfig to toggle the optimized sampling mode for Tensor Parallelism, reducing communication overhead.

- Distributed Greedy Sampling: Implemented a new greedy_sample function that gathers local maximum logits and their global indices across ranks to determine the global argmax, avoiding full vocabulary gathering.

- Distributed Top-K/Top-P Sampling: Enhanced top-k/top-p logic to perform local top-k selection first, then gather candidates across ranks and apply top-p filtering on the combined global candidates.

- Updated Rejection Sampling: Modified core rejection sampling functions (rejection_sample, rejection_random_sample_kernel, etc.) to support compressed vocabulary mode, working with reduced candidate tokens and global indices for efficiency.

- Model-Specific Logits Handling: Patched Eagle3LlamaForCausalLM.compute_logits to return only selected draft token IDs (and bias) when reduce sampling is enabled, optimizing draft generation.

- Pipeline Integration: IntegratedAscendRejectionSampler and AscendTopKTopPSampler into ModelRunnerV1’s sampling pipeline, adding prepare_sampling calls to pass max_topk dynamically.

- vLLM version: v0.20.1

### CODE REVIEW

**Review by weijinqian0 — APPROVED** [review]

**Comment by weijinqian0** [comment]
need avoid to amend related function in reject_sample.

File: vllm_ascend/spec_decode/eagle_proposer.py [hunk 1]
```diff
@@ -926,21 +926,21 @@ def _run_merged_draft(
...
         sample_hidden_states = last_hidden_states[token_indices_to_sample]
-        logits = self.model.compute_logits(sample_hidden_states)
-
-        if lmhead_tp_enable() and num_indices < logits.shape[0]:
-            logits = logits[:num_indices]
-            token_indices_to_sample = token_indices_to_sample[:num_indices]
-
-        draft_token_ids = logits.argmax(dim=-1)
+        logits = self.model.compute_logits(sample_hidden_states, get_ascend_config().enable_reduce_sample)
+        if not get_ascend_config().enable_reduce_sample:
```
Comment (MengqingCao) [inline vllm_ascend/spec_decode/eagle_proposer.py]: I prefer to enable this feature by default, I think this brings a general performance gain.

File: vllm_ascend/spec_decode/eagle_proposer.py [hunk 2]
```diff
@@ -890,7 +891,6 @@ def _run_merged_draft(
         else:
             last_hidden_states, hidden_states = ret_hidden_states
 
-        if self.method != "dflash":
```
Comment (lilinsiman) [inline vllm_ascend/spec_decode/eagle_proposer.py]: Check whether the DFlash needs to be determined.

File: tests/ut/spec_decode/test_eagle_proposer.py [hunk 3]
```diff
@@ -2550,8 +2558,11 @@ def test_run_merged_draft_mtp_mrope_graph_and_lmhead_tp_preparation(self):
         forward_context.attn_metadata = None
         multi_steps_attn_metadata = [MagicMock(), MagicMock(), MagicMock()]
 
+        mock_ascend_config = MagicMock()
+        mock_ascend_config.enable_reduce_sample = False
```
Comment (lilinsiman) [inline tests/ut/spec_decode/test_eagle_proposer.py:2659]: The scenario where reduce_sample is true is added, and assert is modified or added.

File: vllm_ascend/spec_decode/eagle_proposer.py
Comment (lilinsiman) [inline vllm_ascend/spec_decode/eagle_proposer.py]: Check whether reduce_sample and lmhead can be enabled at the same time. The logic of the judgment in this line is inconsistent with that in line 930.

**Review by linfeng-yuan — APPROVED** [review]
LGTM.
with 3 questions:
1) add an e2e case with mtp and sampling metdata
2) answer Jinqian's question about implementation of rej_sampler
3) update the reason of patch and elimination plan in `vllm_ascend/patch/__init__.py`

File: vllm_ascend/spec_decode/eagle_proposer.py (hunk 1 above)
Comment (hzx55906) [reply vllm_ascend/spec_decode/eagle_proposer.py]: I have set this optimization to be enabled by default, so no manual activation is required in the settings.

File: vllm_ascend/spec_decode/eagle_proposer.py (hunk 2 above)
Comment (hzx55906) [reply vllm_ascend/spec_decode/eagle_proposer.py]: Fixed. This optimization also applies to DFlash.

File: tests/ut/spec_decode/test_eagle_proposer.py (hunk 3 above)
Comment (hzx55906) [reply tests/ut/spec_decode/test_eagle_proposer.py:2659]: After enabling this optimization by default, the issue of incomplete UT test coverage has been resolved.

File: vllm_ascend/spec_decode/eagle_proposer.py
Comment (hzx55906) [reply vllm_ascend/spec_decode/eagle_proposer.py]: Fixed. This optimization can still be enabled in the lmhead_tp_enable scenario by simply delaying the all-to-all communication following the lmhead layer.

File: vllm_ascend/sample/rejection_sampler.py [hunk 4]
```diff
@@ -18,34 +25,211 @@
     rejection_random_sample_kernel,
     sample_recovered_tokens_kernel,
 )
+from vllm_ascend.sample.penalties import apply_all_penalties
 from vllm_ascend.sample.sampler import apply_top_k_top_p
 
 
+class AscendRejectionSampler(RejectionSampler):
```
Comment (wangxiyuan) [inline vllm_ascend/sample/rejection_sampler.py:33]: this is back again, any idea? @realliujiaxu

**Comment by MengqingCao** [comment]
plz rebase to the latest code to address the CI issue

**Review by MengqingCao — APPROVED** [review]

File: vllm_ascend/ops/triton/reject_sample.py [hunk 5]
```diff
@@ -145,16 +145,20 @@ def rejection_random_sample_kernel(
...
+    target_probs_ptr,  # [num_tokens, vocab_size] or [num_tokens, selected_vocab_size] if ENABLE_REDUCE_SAMPLING
+    target_indices_ptr,  # [num_tokens, selected_vocab_size] global vocab indices, only used if ENABLE_REDUCE_SAMPLING
     bonus_token_ids_ptr,  # [batch_size]
     recovered_token_ids_ptr,  # [num_tokens]
     uniform_probs_ptr,  # [num_tokens]
     is_greedy_ptr,  # [batch_size]
     max_spec_len,
-    vocab_size,
+    vocab_size,  # vocab_size or selected_vocab_size if ENABLE_REDUCE_SAMPLING
+    global_vocab_size,  # global vocab size for draft_probs indexing (only used if ENABLE_REDUCE_SAMPLING)
```
Comment (MengqingCao) [inline vllm_ascend/ops/triton/reject_sample.py:156]: it seems the suggestions from @weijinqian0 has not been fixed now?
